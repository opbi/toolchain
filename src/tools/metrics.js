import prometheus from 'prom-client';

import metricsClient from './metrics-client';

// TODO: update the original .inc(), .label() function
// to take only the labelNames[] from the input

// TODO: further abstract out type of metrics will be used on application
// e.g. error_counter, call_timer, etc.
// make metrics-specs to { action, type }?

// TODO: consider using fastify-metrics to get default histogram of response times

// TODO: need to test the behaviour in a distributed system

let metricsUnits = {};

/*
  when metrics were scraped by prometheus deamon,
  k8s meta data such as `app` would be added
  so it is not very necessary to add AppName into metrics namespace
  it can be used to distinguish with canned metrics
 */
const metricsNamespace = process.env.METRICS_NAMESPACE || 'custom_metrics';

/**
 * Add a new metrics unit from metrics-specs.
 *
 * @param {object} options - Config.
 * @param  {string}    options.type -  Type of the metrics, e.g. Counter.
 * @param  {string}    options.name -  Name of the metrics, e.g. Action_error.
 * @param  {object} options.specs - Other specific configuration for the metrics, e.g. Help, labelNames, buckets.
 */
const add = ({ type, name, ...specs }) => {
  const MetricsConstructor = metricsClient[type];

  if (!MetricsConstructor) {
    throw Error('invalid metrics type');
  }

  // metricsNamespace will be appended
  const m = new MetricsConstructor({
    ...specs,
    name: `${metricsNamespace}_${name}`,
  });

  metricsUnits[name] = m;
};

/**
 * Load a list of metrics specs to metricsUnits.
 *
 * @param  {Array} list - List of metrics specs.
 */
const load = (list) => {
  list.forEach((m) => {
    add(m);
  });
};

/**
 * List the metrics unit instances stored in metricsUnits.
 *
 * @returns {object} <{[metrics name]: metrics unit instance}>.
 */
const list = () => metricsUnits;

/**
 * Find a metrics unit by action/type or by name(higher priority).
 *
 * @param {object} options - Config.
 * @param  {string} options.action - Name of the action function.
 * @param  {string} options.type -   Type of metrics attached to the action function.
 * @param  {string} options.name -   Name of the metrics unit.
 * @returns {object}                The metrics unit instance.
 */
const find = ({ action, type, name }) => {
  const unitName = name || `${action}_${type}`;
  const unit = metricsUnits[unitName];
  if (!unit) {
    throw Error(`${unitName} not found in metrics-specs`);
  }
  return unit;
};

/**
 * Clear instances stored in metricsUnits and prometheus.register.
 */
const reset = () => {
  prometheus.register.clear();
  metricsUnits = {};
};

export default {
  add,
  load,
  list,
  find,
  reset,
};
