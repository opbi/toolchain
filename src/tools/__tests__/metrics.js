import metrics from '../metrics';

const metricsSpecs = [
  {
    name: 'serverRestart',
    type: 'counter',
    help: `the number of times server(node instance) restarted, an unexpected high number indicates crash loop`,
  },
  {
    name: 'dbConnect_timer',
    type: 'timer',
    help: `the time to create db connection to monitor upstream service health`,
    labelNames: ['envId'],
  },
  {
    name: 'dbConnect_error',
    type: 'counter',
    help: `the number of db connection errors`,
    labelNames: ['envId'],
  },
  {
    name: 'orderValue',
    type: 'histogram',
    help: 'price summary of shopping orders',
    buckets: [10, 100, 1000, 10000, 100000],
    labelNames: ['envId', 'length'],
  },
];

describe('metrics', () => {
  beforeEach(() => {
    metrics.load(metricsSpecs);
  });

  afterEach(() => {
    metrics.reset();
  });

  it('is an object with the following methods', () => {
    expect(typeof metrics).toBe('object');
    expect(metrics).toHaveProperty('load');
    expect(metrics).toHaveProperty('add');
    expect(metrics).toHaveProperty('find');
  });

  it('.load() can create a list of metrics unit', () => {
    metrics.reset();
    metrics.load(metricsSpecs);
    const units = metrics.list();
    expect(Object.keys(units)).toHaveLength(4);
  });

  it('.load() throws an error if invalid metrics type was specified', () => {
    metrics.reset();
    const invalidSpecs = [
      {
        name: 'someName',
        type: 'randomType',
      },
    ];
    try {
      metrics.load(invalidSpecs);
    } catch (e) {
      expect(e).toBeDefined();
    }
  });

  it('.find() can find a metrics unit', () => {
    const dbConnectTimer = metrics.find({ action: 'dbConnect', type: 'timer' });
    expect(dbConnectTimer).toBeDefined();
  });

  it('.find() throws an error if no unit found', () => {
    try {
      metrics.find({ action: 'dbConnect', type: 'invalid' });
    } catch (e) {
      expect(e).toBeDefined();
    }
  });

  it('counter.count() with excessive labels', () => {
    const counter = metrics.find({ name: 'dbConnect_error' });
    counter.count({ envId: 'test', foo: 'bar' });
    counter.count({ envId: 'test', foo: 'bar' }, 2);
    expect(counter.value()).toBe(3);
  });

  it('counter.value() to get the current total count with optional filter', () => {
    const counter = metrics.find({ name: 'dbConnect_error' });
    expect(counter.value()).toBe(0);
    counter.count();
    expect(counter.value()).toBe(1);
    counter.count({ envId: 'test' });
    counter.count({ envId: 'prod' });
    expect(counter.value({ envId: 'test' })).toBe(1);
    expect(counter.value({ envId: 'prod' })).toBe(1);
  });

  it('counter.stats() to get distribution of all values of a label', () => {
    const counter = metrics.find({ name: 'dbConnect_error' });
    expect(counter.value()).toBe(0);
    counter.count();
    counter.count({ envId: 'test' });
    counter.count({ envId: 'test' });
    counter.count({ envId: 'prod' });
    expect(counter.stats('envId')).toMatchSnapshot();
  });

  it('timer.start() with excessive labels or none', async () => {
    const timer = metrics.find({ name: 'dbConnect_timer' });
    const timerStop = timer.start({ envId: 'test', foo: 'bar' });
    timerStop();
    timer.start()();
    expect(timer.get().values[0]).toBeDefined();
  });
});
