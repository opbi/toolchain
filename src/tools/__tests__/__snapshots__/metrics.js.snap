// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`metrics counter.stats() to get distribution of all values of a label 1`] = `
Object {
  "prod": 1,
  "test": 2,
  "undefined": 1,
}
`;
